static/app/components/forms/compactSelect.tsx

```diff
@@ -178,41 +178,51 @@ function CompactSelect<OptionType extends GeneralSelectValue = GeneralSelectValu
   // Keep an internal copy of the current select value and update the control
   // button's label when the value changes
   const [internalValue, setInternalValue] = useState(valueProp ?? defaultValue);
-  const [label, setLabel] = useState(getLabel(valueProp ?? null));
 
   // Update the button label when the value changes
-  function getLabel(newValue): React.ReactNode {
+  const getLabel = useCallback((): React.ReactNode => {
+    const newValue = valueProp ?? internalValue;
     const valueSet = Array.isArray(newValue) ? newValue : [newValue];
-    const optionSet = valueSet
+    const selectedOptions = valueSet
       .map(val => getSelectedOptions<OptionType>(options, val))
       .flat();
-    const firstOptionLabel = optionSet[0]?.label ?? '';
 
     return (
       <Fragment>
-        <ButtonLabel>{firstOptionLabel}</ButtonLabel>
-        {optionSet.length > 1 && <StyledBadge text={`+${optionSet.length - 1}`} />}
+        <ButtonLabel>{selectedOptions[0]?.label ?? ''}</ButtonLabel>
+        {selectedOptions.length > 1 && (
+          <StyledBadge text={`+${selectedOptions.length - 1}`} />
+        )}
       </Fragment>
     );
-  }
+  }, [options, valueProp, internalValue]);
```

Comment:
Wrapping `getLabel()` inside a `useCallback()`, so it doesn't get re-defined on every render.